### ISSUE TITLE
JsonSerializer NRE when deserializing object from array (bad error message)

### ISSUE THREAD

**Issue by zivkan** [dotnet/corefx#42143]
I accidentally tried to deserialize some JSON using the wrong datatype and got a `NullReferenceException`. I think the error message should be improved.

sample:
```csharp
using System;
using System.Collections.Generic;

namespace stjCrash
{
    class Program
    {
        static void Main(string[] args)
        {
            var json = "{\"key\":[{\"name\": \"something\"}]}";

            TryRun("System.Text.Json correct type", () => System.Text.Json.JsonSerializer.Deserialize<Dictionary<string, List<Value>>>(json));
            TryRun("System.Text.Json wrong type", () => System.Text.Json.JsonSerializer.Deserialize<Dictionary<string, Value>>(json));

            TryRun("Newtonsoft.Json correct type", () => Newtonsoft.Json.JsonConvert.DeserializeObject<Dictionary<string, List<Value>>>(json));
            TryRun("Newtonsoft.Json wrong type", () => Newtonsoft.Json.JsonConvert.DeserializeObject<Dictionary<string, Value>>(json));
        }

        private static void TryRun(string name, Func<object> a)
        {
            try
            {
                a.Invoke();
                Console.WriteLine(name + " succeeded");
            }
            catch (Exception e)
            {
                Console.WriteLine(name + " crashed (" + e.GetType().Name + "): " + e.Message);
            }
        }

        private class Value
        {
            public string Name { get; set; }
        }
    }
}
```

output:
```text
System.Text.Json correct type succeeded
System.Text.Json wrong type crashed (NullReferenceException): Object reference not set to an instance of an object.
Newtonsoft.Json correct type succeeded
Newtonsoft.Json wrong type crashed (JsonSerializationException): Cannot deserialize the current JSON array (e.g. [1,2,3]) into type 'stjCrash.Program+Value' because the type requires a JSON object (e.g. {"name":"value"}) to deserialize correctly.
To fix this error either change the JSON to a JSON object (e.g. {"name":"value"}) or change the deserialized type to an array or a type that implements a collection interface (e.g. ICollection, IList) like List<T> that can be deserialized from a JSON array. JsonArrayAttribute can also be added to the type to force it to deserialize from a JSON array.
Path 'key', line 1, position 8.
```

**Comment by ahsonkhan** [comment]
A similar issue (where object was being deserialized as array) was fixed recently: https://github.com/dotnet/corefx/issues/41839

This issue is the reverse case, where an array is incorrectly being deserialized as object.

Are you seeing this issue in master, using the latest nightly build? To try that, you could grab the latest runtime (see https://github.com/dotnet/core-setup#daily-builds), or use the latest `System.Text.Json` NuGet package available from this feed (version `5.0.0-...`):
https://dotnetfeed.blob.core.windows.net/dotnet-core/index.json

There are several test cases here, so if you have observed a case missing, please add it (along with variants related to it), here:
https://github.com/dotnet/corefx/blob/bc78f93494ea732a5852ce46aded6ecfc5135731/src/System.Text.Json/tests/Serialization/DictionaryTests.cs#L381-L394

**Comment by zivkan** [comment]
> Are you seeing this issue in master, using the latest nightly build?

yes. When I created the test before the code fix in my PR, I was getting the NRE, so I'm confident it's a bug still in master.

I've found a simplified test that fails, `.Deserialize<int>(@"[1]")`, so I'm not sure that `DictionaryTests.cs` is the best place, since it doesn't need a dictionary. I'll take a look if I can find a better place for the new tests, but let me know if you have another suggestion.
